pgx/src/pgbox.rs

```diff
@@ -229,54 +254,25 @@ impl<T> PgBox<T> {
     }
 
     /// Box nothing
-    pub fn null() -> PgBox<T> {
-        PgBox::<T> {
-            inner: Inner::<T> {
-                ptr: None,
-                allocated_by_pg: false,
-            },
-        }
-    }
-
-    /// Box a pointer that came from Postgres.
-    ///
-    /// When this `PgBox<T>` is dropped, the boxed memory is **not** freed.  Since Postgres
-    /// allocated it, Postgres is responsible for freeing it.
     #[inline]
-    pub unsafe fn from_pg(ptr: *mut T) -> PgBox<T> {
-        PgBox::<T> {
-            inner: Inner::<T> {
-                ptr: if ptr.is_null() { None } else { Some(ptr) },
-                allocated_by_pg: true,
-            },
-        }
-    }
-
-    /// Box a pointer that was allocated within Rust
-    ///
-    /// When this `PgBox<T>` is dropped, the boxed memory is freed.  Since Rust
-    /// allocated it, Rust is responsible for freeing it.
-    ///
-    /// If you need to give the boxed pointer to Postgres, call `.into_pg()`
-    pub unsafe fn from_rust(ptr: *mut T) -> PgBox<T> {
-        PgBox::<T> {
-            inner: Inner {
-                ptr: if ptr.is_null() { None } else { Some(ptr) },
-                allocated_by_pg: false,
-            },
+    pub fn null() -> PgBox<T, AllocatedByPostgres> {
+        PgBox::<T, AllocatedByPostgres> {
+            ptr: None,
+            __marker: PhantomData,
         }
     }
 
     /// Are we boxing a NULL?
+    #[inline]
     pub fn is_null(&self) -> bool {
-        self.inner.ptr.is_none()
+        self.ptr.is_none()
```

Comment:
Should be impossible to have a null pointer within the `Some` that we've created, but ...

```suggestion
        match self.ptr {
            Some(ptr) => ptr.is_null(),
            None => true,
        }
```

Reply:
Not exactly:  https://github.com/zombodb/pgx/blob/2b62ea2ee5905ebf1c638beccecda7dbc5ccc1a9/pgx/src/pgbox.rs#L258